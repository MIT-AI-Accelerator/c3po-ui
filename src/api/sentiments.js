/**
   * Sends a prompt to the GPT-J model for a seq2seq prediction.
   * @param {string} text
   */
export async function submitPrompt(
    text = "Placeholder text here."
) {
    const apiResp = await fetch(`http://127.0.0.1:8000/v1/sentiments/insights/getchatstress`, {
        method: 'POST',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify({
            text
        }),
    });
    const respJson = await apiResp.json();
    return {
        status: apiResp.status,
        data: respJson,
    };
}



// async function submitPrompt(
//     text="Placeholder text here."
// ) {
//     const apiResp = await fetch(`${this.REACT_APP_COMMON_API_BASE}/scratch`, {
//         method: 'POST',
//         headers: {
//             'content-type': 'application/json',
//         },
//         credentials: 'include',
//         body: JSON.stringify({
//             appId: this.REACT_APP_APPID,
//             key,
//             value,
//         }),
//     });
//     const respJson = await apiResp.json();
//     return {
//         status: apiResp.status,
//         data: respJson,
//     };
// }

