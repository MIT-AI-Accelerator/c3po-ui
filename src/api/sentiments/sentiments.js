/**
   * Sends a prompt to the GPT-J model for a seq2seq prediction.
   * @param {string} text
   */
export async function submitPrompt(
    text = "Placeholder text here."
) {
    const apiResp = await fetch(`${process.env.REACT_APP_API_BASE_URL}/v1/sentiments/insights/getchatstress`, {
        method: 'POST',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify({
            text
        }),
    });

    const respJson = await apiResp.json();
    return {
        status: apiResp.status,
        data: respJson,
    };
}