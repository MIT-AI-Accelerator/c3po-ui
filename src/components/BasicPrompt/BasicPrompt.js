import React, { useState, useEffect } from 'react';
import OutlinedInput from '@mui/material/OutlinedInput';
import InputLabel from '@mui/material/InputLabel';
import InputAdornment from '@mui/material/InputAdornment';
import FormControl from '@mui/material/FormControl';
import Paper from '@mui/material/Paper';

import { submitPrompt } from '../../api/sentiments/sentiments';

export function BasicPrompt() {
  const [answer, setAnswer] = useState(null);
  const [prompt, setPrompt] = useState("");

  useEffect(() => {

    async function getAnswer(prompt) {
      const newAnswerObj = await submitPrompt(prompt);
      newAnswerObj.status === 200 && setAnswer(newAnswerObj.data.answer);
      newAnswerObj.status !== 200 && setAnswer('Error, try again later.');
    }

    if (prompt) {
      setAnswer("");
      getAnswer(prompt);
    }

  }, [prompt]);


  const handlePromptEnter = (event) => {
    if (event.key === "Enter") {
      setPrompt(event.target.value);
    }
  };

  return (
      <Paper
        sx={{
          p: 2,
          display: 'flex',
          flexDirection: 'column',
          height: 240,
        }}
      >
        <div>
          <FormControl fullWidth sx={{ ml: 0, mr: 1 }}>
            <InputLabel htmlFor="outlined-adornment-amount">Stress Gauge</InputLabel>
            <OutlinedInput
              id="outlined-adornment-amount"
              onKeyUp={handlePromptEnter}
              startAdornment={<InputAdornment position="start">{">>>"}</InputAdornment>}
              label="C3PO Prompt"
            />
          </FormControl>
          {answer === null && <p>
            <code>Type a message to view stress level</code>
          </p>}
          {answer === "" && <p>
            <code>Loading...</code>
          </p>}
          {answer && <p>
            <code>{answer}</code>
          </p>}
        </div>
      </Paper>
  );
}




