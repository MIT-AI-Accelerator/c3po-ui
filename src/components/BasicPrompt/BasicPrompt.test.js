import { render, screen, within, cleanup, fireEvent } from '@testing-library/react';
import { BasicPrompt } from './BasicPrompt';
import * as sentimentsApi from "../../api/sentiments/sentiments";

afterEach(cleanup);

describe('App', () => {
  it('renders App component', () => {
    render(<BasicPrompt />);
  });

  it('Contains an input with a Stress Gauge label', () => {
    render(<BasicPrompt />);
    const labelElement = within(document.getElementsByTagName("fieldset")[0]).getByText(/C3PO/);
    const inputElement = screen.getByRole("textbox");

    expect(labelElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();

  });

  it('Matches the last snapshot', () => {
    const { asFragment } = render(<BasicPrompt />);
    expect(asFragment()).toMatchSnapshot();
  });

});

describe('Interactions with main text input', () => {

  beforeEach(() => {
    const submitPromptMock = jest.spyOn(sentimentsApi, "submitPrompt");

    submitPromptMock.mockImplementation(async (prompt) => {
      return prompt === 'good' ?
        { status: 200, data: { answer: "Great Success" } } :
        { status: 400 }
    });
  })

  afterEach(() => {
    const submitPromptMock = jest.spyOn(sentimentsApi, "submitPrompt");
    submitPromptMock.mockReset();
  });

  it('...calls the API when the prompt updates and returns the response when 200', async () => {

    // render app, save the react-testing-library utils
    const { queryByLabelText, getByLabelText, findByText } = render(
      <BasicPrompt />,
    );

    // ensure that the form input exists
    expect(queryByLabelText(/Stress Gauge/i)).toBeTruthy();

    // "type" the word "good" and hit enter
    fireEvent.keyUp(getByLabelText(/Stress Gauge/i), { key: 'Enter', code: 'Enter', charCode: 13, target: { value: "good" } });

    // ensure the correct api function was called with the correct word
    expect(sentimentsApi.submitPrompt).toHaveBeenCalledWith('good');

    // since the api is async (even if mocked), wait for the state to update using findBy (or could waitFor)
    expect(await findByText(/Great Success/i)).toBeTruthy();
  });

  it('...calls the API when the prompt updates and tells the user of error when not 200', async () => {

    // render app, save the react-testing-library utils
    const { queryByLabelText, getByLabelText, findByText } = render(
      <BasicPrompt />,
    );

    // ensure that the form input exists
    expect(queryByLabelText(/Stress Gauge/i)).toBeTruthy();

    // "type" the word "bad" and hit enter
    fireEvent.keyUp(getByLabelText(/Stress Gauge/i), { key: 'Enter', code: 'Enter', charCode: 13, target: { value: "bad" } });

    // ensure the correct api function was called with the correct word
    expect(sentimentsApi.submitPrompt).toHaveBeenCalledWith('bad');

    // since the api is async (even if mocked), wait for the state to update using findBy (or could waitFor)
    expect(await findByText(/Error, try again later./i)).toBeTruthy();
  });

  it('...not to call API if not Enter', async () => {

    // render app, save the react-testing-library utils
    const { queryByLabelText, getByLabelText, findByText } = render(
      <BasicPrompt />,
    );

    // ensure that the form input exists
    expect(queryByLabelText(/Stress Gauge/i)).toBeTruthy();

    // "type" the key d
    fireEvent.keyUp(getByLabelText(/Stress Gauge/i), { key: 'd', code: 'KeyD', charCode: 68, target: { value: "anything" } });

    // not called
    expect(sentimentsApi.submitPrompt).not.toBeCalled();

  });

});