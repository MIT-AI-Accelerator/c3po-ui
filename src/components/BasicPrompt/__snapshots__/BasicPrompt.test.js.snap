// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Matches the last snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-t5ouwv-MuiPaper-root"
  >
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-uvw8af-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="outlined-adornment-amount"
        >
          Stress Gauge
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              &gt;&gt;&gt;
            </p>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
            id="outlined-adornment-amount"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                C3PO Prompt
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <p>
        <code>
          Type a message to view stress level
        </code>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
